File: packages/polygon-packer/src/genetic-algorithm/genetic-algorithm.ts
import { abs_polygon_area, rotate_polygon_wasm, calculate_bounds_wasm } from 'wasm-nesting';
import { BoundRectF32 } from '../geometry';
import { NestConfig, PolygonNode } from '../types';
import Phenotype from './phenotype';
export default class GeneticAlgorithm {
    #binBounds: BoundRectF32 = new BoundRectF32();

    #population: Phenotype[] = [];

    #rotations: number = 0;

    #trashold: number = 0;

    public init(nodes: PolygonNode[], bounds: BoundRectF32, config: NestConfig): void {
        this.#rotations = config.rotations;
        this.#trashold = 0.01 * config.mutationRate;
        this.#binBounds.from(bounds);

        // initiate new GA
        const adam: PolygonNode[] = nodes.slice();

        adam.sort((a, b) => abs_polygon_area(b.memSeg) - abs_polygon_area(a.memSeg));
        // population is an array of individuals. Each individual is a object representing the
        // order of insertion and the angle each part is rotated
        const angles: number[] = [];
        let mutant: Phenotype = null;

        for (let i = 0; i < adam.length; ++i) {
            angles.push(this.randomAngle(adam[i]));
        }

        this.#population.push(new Phenotype(adam.map(node => node.source), angles));

        while (this.#population.length < config.populationSize) {
            mutant = this.mutate(nodes, this.#population[0]);
            this.#population.push(mutant);
        }
    }

    public clean(): void {
        this.#rotations = 0;
        this.#trashold = 0;
        this.#binBounds.clean();
        this.#population.length = 0;
    }

    // returns a mutated individual with the given mutation rate
    private mutate(nodes: PolygonNode[], individual: Phenotype): Phenotype {
        const clone: Phenotype = individual.clone();
        const size: number = clone.size;
        let i: number = 0;

        for (i = 0; i < size; ++i) {
            if (this.getMutate()) {
                clone.swap(i);
            }

            if (this.getMutate()) {
                clone.rotation[i] = this.randomAngle(nodes[clone.placement[i]]);
            }
        }

        return clone;
    }

    // single point crossover
    private mate(male: Phenotype, female: Phenotype): Phenotype[] {
        const cutPoint: number = male.cutPoint;
        const result: Phenotype[] = [male.cut(cutPoint), female.cut(cutPoint)];

        result[0].mate(female);
        result[1].mate(male);

        return result;
    }

    // returns a random individual from the population, weighted to the front of the list (lower
    // fitness value is more likely to be selected)
    private randomWeightedIndividual(exclude?: Phenotype): Phenotype {
        const excludeIndex: number = exclude ? this.#population.indexOf(exclude) : -1;
        const localPopulation: Phenotype[] = this.#population.slice();

        if (excludeIndex !== -1) {
            localPopulation.splice(excludeIndex, 1);
        }

        const size: number = localPopulation.length;
        const rand: number = Math.random();
        const weight: number = 2 / size;
        let lower: number = 0;
        let upper: number = weight / 2;
        let i: number = 0;

        for (i = 0; i < size; ++i) {
            // if the random number falls between lower and upper bounds, select this individual
            if (rand > lower && rand < upper) {
                return localPopulation[i];
            }

            lower = upper;
            upper = upper + weight * ((size - i) / size);
        }

        return localPopulation[0];
    }

    // returns a random angle of insertion
    private randomAngle(node: PolygonNode): number {
        const lastIndex: number = this.#rotations - 1;
        const angles: number[] = [];
        const step: number = 360 / this.#rotations;
        let angle: number = 0;
        let i: number = 0;
        let j: number = 0;

        for (i = 0; i < this.#rotations; ++i) {
            angles.push(Math.round(i * step));
        }

        for (i = lastIndex; i > 0; --i) {
            j = Math.floor(Math.random() * (i + 1));
            angle = angles[i];
            angles[i] = angles[j];
            angles[j] = angle;
        }

        for (i = 0; i < this.#rotations; ++i) {
            const rotated = rotate_polygon_wasm(node.memSeg, angles[i]);
            const bounds = calculate_bounds_wasm(rotated);

            // don't use obviously bad angles where the part doesn't fit in the bin
            if (bounds[2] < this.#binBounds.width && bounds[3] < this.#binBounds.height) {
                return angles[i];
            }
        }

        return 0;
    }

    public getIndividual(nodes: PolygonNode[]): Phenotype {
        // evaluate all members of the population
        for (let i = 0; i < this.#population.length; ++i) {
            if (!this.#population[i].fitness) {
                return this.#population[i];
            }
        }

        // all individuals have been evaluated, start next generation
        // Individuals with higher fitness are more likely to be selected for mating
        this.#population.sort((a, b) => a.fitness - b.fitness);

        // fittest individual is preserved in the new generation (elitism)
        const result: Phenotype[] = [this.#population[0]];

        while (result.length < this.#population.length) {
            const male = this.randomWeightedIndividual();
            const female = this.randomWeightedIndividual(male);

            // each mating produces two children
            const children = this.mate(male, female);

            // slightly mutate children
            result.push(this.mutate(nodes, children[0]));

            if (result.length < this.#population.length) {
                result.push(this.mutate(nodes, children[1]));
            }
        }

        this.#population = result;

        return this.#population[1];
    }

    private getMutate(): boolean {
        return Math.random() < this.#trashold;
    }
}

File: packages/wasm-nesting/src/genetic_algorithm/genetic_algorithm.rs
use super::phenotype::Phenotype;
use crate::{
    geometry::bound_rect::BoundRect, nest_config::NestConfig, nesting::polygon_node::PolygonNode,
    utils::number::Number,
};
use rand::Rng;

pub struct GeneticAlgorithm {
    bin_width: f32,
    bin_height: f32,
    population: Vec<Phenotype>,
    rotations: u8,
    threshold: f32,
    current_source: u16,
}

impl GeneticAlgorithm {
    pub fn new() -> Self {
        GeneticAlgorithm {
            bin_width: 0.0,
            bin_height: 0.0,
            population: Vec::new(),
            rotations: 0,
            threshold: 0.0,
            current_source: 0,
        }
    }

    pub fn init(&mut self, nodes: &[PolygonNode], bounds: &BoundRect<f32>, config: &NestConfig) {
        self.rotations = config.rotations;
        self.threshold = 0.01 * config.mutation_rate as f32;
        self.bin_width = unsafe { bounds.width() };
        self.bin_height = unsafe { bounds.height() };

        // Initiate new GA
        let mut adam: Vec<PolygonNode> = nodes.to_vec();

        // Sort by area (largest first)
        adam.sort_by(|a, b| {
            let area_a = f32::polygon_area(&a.mem_seg).abs();
            let area_b = f32::polygon_area(&b.mem_seg).abs();
            area_b
                .partial_cmp(&area_a)
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        // Create initial individual
        let sources: Vec<i32> = adam.iter().map(|node| node.source).collect();
        let mut angles: Vec<u16> = Vec::with_capacity(adam.len());

        for node in &adam {
            angles.push(self.random_angle(node));
        }

        self.population
            .push(Phenotype::new(self.current_source, sources, angles));
        self.current_source += 1;

        // Create population through mutation
        let first = self.population[0].clone_with_source(self.population[0].source());
        while self.population.len() < config.population_size as usize {
            let mutant = self.mutate(&adam, &first);
            self.population.push(mutant);
        }
    }

    pub fn clean(&mut self) {
        self.rotations = 0;
        self.threshold = 0.0;
        self.bin_width = 0.0;
        self.bin_height = 0.0;
        self.population.clear();
        self.current_source = 0;
    }

    // Returns a mutated individual with the given mutation rate
    fn mutate(&mut self, nodes: &[PolygonNode], individual: &Phenotype) -> Phenotype {
        let mut clone = individual.clone_with_source(self.current_source);
        self.current_source += 1;

        let size = clone.size();

        for i in 0..size {
            if self.get_mutate() {
                clone.swap(i);
            }

            if self.get_mutate() {
                let placement_idx = clone.placement()[i] as usize;
                clone.rotation_mut()[i] = self.random_angle(&nodes[placement_idx]);
            }
        }

        clone
    }

    // Single point crossover
    fn mate(&mut self, male: &Phenotype, female: &Phenotype) -> [Phenotype; 2] {
        let cut_point = male.cut_point();
        let mut child1 = male.cut(self.current_source, cut_point);
        let mut child2 = female.cut(self.current_source + 1, cut_point);

        self.current_source += 2;

        child1.mate(female);
        child2.mate(male);

        [child1, child2]
    }

    // Returns a random individual from the population, weighted to the front of the list
    // (lower fitness value is more likely to be selected)
    fn random_weighted_individual(&self, exclude: Option<usize>) -> usize {
        let mut local_indices: Vec<usize> = (0..self.population.len()).collect();

        if let Some(exclude_idx) = exclude {
            local_indices.retain(|&idx| idx != exclude_idx);
        }

        let size = local_indices.len();
        let mut rng = rand::thread_rng();
        let rand: f32 = rng.gen();
        let weight = 2.0 / size as f32;
        let mut lower = 0.0;
        let mut upper = weight / 2.0;

        for i in 0..size {
            // If the random number falls between lower and upper bounds, select this individual
            if rand > lower && rand < upper {
                return local_indices[i];
            }

            lower = upper;
            upper = upper + weight * ((size - i) as f32 / size as f32);
        }

        local_indices[0]
    }

    // Returns a random angle of insertion
    fn random_angle(&self, node: &PolygonNode) -> u16 {
        let last_index = self.rotations as usize - 1;
        let step = 360.0 / self.rotations as f32;
        let mut angles: Vec<u16> = Vec::with_capacity(self.rotations as usize);

        for i in 0..self.rotations as usize {
            angles.push((i as f32 * step).round() as u16);
        }

        // Shuffle angles
        let mut rng = rand::thread_rng();
        for i in (1..=last_index).rev() {
            let j = rng.gen_range(0..=i);
            angles.swap(i, j);
        }

        // Try each angle and find one that fits
        for angle in &angles {
            let mut rotated = node.mem_seg.to_vec();
            f32::rotate_polygon(&mut rotated, *angle as f32);

            let size = rotated.len() >> 1;
            let bounds = f32::calculate_bounds(&rotated, 0, size);

            // Don't use obviously bad angles where the part doesn't fit in the bin
            if bounds[2] < self.bin_width && bounds[3] < self.bin_height {
                return *angle;
            }
        }

        0
    }

    pub fn get_individual(&mut self, nodes: &[PolygonNode]) -> Option<&Phenotype> {
        // Check if all members have been evaluated
        let all_evaluated = self.population.iter().all(|p| p.fitness() > 0.0);

        if !all_evaluated {
            // Find first unevaluated individual
            for i in 0..self.population.len() {
                if self.population[i].fitness() == 0.0 {
                    return Some(&self.population[i]);
                }
            }
        }

        // All individuals have been evaluated, start next generation
        // Sort by fitness (lower is better)
        self.population.sort_by(|a, b| {
            a.fitness()
                .partial_cmp(&b.fitness())
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        // Fittest individual is preserved in the new generation (elitism)
        let mut result = vec![self.population[0].clone_with_source(self.population[0].source())];

        while result.len() < self.population.len() {
            let male_idx = self.random_weighted_individual(None);
            let female_idx = self.random_weighted_individual(Some(male_idx));

            let male =
                self.population[male_idx].clone_with_source(self.population[male_idx].source());
            let female =
                self.population[female_idx].clone_with_source(self.population[female_idx].source());

            // Each mating produces two children
            let children = self.mate(&male, &female);

            // Slightly mutate children
            result.push(self.mutate(nodes, &children[0]));

            if result.len() < self.population.len() {
                result.push(self.mutate(nodes, &children[1]));
            }
        }

        self.population = result;

        if self.population.len() > 1 {
            Some(&self.population[1])
        } else {
            None
        }
    }

    pub fn update_fitness(&mut self, source: u16, fitness: f32) {
        for phenotype in &mut self.population {
            if phenotype.source() == source {
                phenotype.set_fitness(fitness);
                break;
            }
        }
    }

    fn get_mutate(&self) -> bool {
        let mut rng = rand::thread_rng();
        let random_value: f32 = rng.gen();
        random_value < self.threshold
    }
}
